pkg/runtime/framework/plugins/mpi/mpi.go

```diff
@@ -176,90 +216,96 @@ func (m *MPI) ReconcilerBuilders() []runtime.ReconcilerBuilder {
 	if info == nil || info.RuntimePolicy.MLPolicy == nil || info.RuntimePolicy.MLPolicy.MPI == nil {
 		return nil, nil
 	}
 
-	secret, err := m.buildSSHAuthSecret(trainJob)
-	if err != nil {
-		return nil, fmt.Errorf("failed to build Secret with SSH auth keys. Error: %v", err)
-	}
+	var objects []any
 
+	// SSHAuthSecret is immutable.
+	if err := m.client.Get(ctx, client.ObjectKey{Name: sshAuthSecretName(trainJob.Name), Namespace: trainJob.Namespace}, &corev1.Secret{}); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return nil, err
+		}
+		secret, err := m.buildSSHAuthSecret(trainJob)
+		if err != nil {
+			return nil, fmt.Errorf("%w: %w", errorFailedToBuildSecret, err)
+		}
+		objects = append(objects, secret)
+	}
 	configMap, err := m.buildHostFileConfigMap(info, trainJob)
 	if err != nil {
-		return nil, fmt.Errorf("failed to build ConfigMap with hostfile. Error: %v", err)
+		return nil, fmt.Errorf("%w: %w", errorFailedToBuildConfigMap, err)
 	}
-
-	return []any{secret, configMap}, nil
+	return append(objects, configMap), nil
 }
 
 func (m *MPI) buildSSHAuthSecret(trainJob *trainer.TrainJob) (*corev1ac.SecretApplyConfiguration, error) {
-	// Generate SSH private and public keys.
 	privateKey, err := ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
 	if err != nil {
-		return nil, fmt.Errorf("failed to generate private SSH key, Error: %v", err)
+		return nil, err
 	}
-
 	privateDER, err := x509.MarshalECPrivateKey(privateKey)
 	if err != nil {
-		return nil, fmt.Errorf("failed to covert private SSH key to DER format. Error: %v", err)
+		return nil, err
 	}
-
 	privatePEM := pem.EncodeToMemory(&pem.Block{
 		Type:  "EC PRIVATE KEY",
 		Bytes: privateDER,
 	})
-
 	publicKey, err := ssh.NewPublicKey(&privateKey.PublicKey)
 	if err != nil {
-		return nil, fmt.Errorf("failed to generate public SSH key. Error:  %v", err)
+		return nil, err
 	}
-
-	// Create Secret to store ssh keys.
-	secret := corev1ac.Secret(trainJob.Name+constants.MPISSHAuthSecretSuffix, trainJob.Namespace).
+	return corev1ac.Secret(sshAuthSecretName(trainJob.Name), trainJob.Namespace).
 		WithType(corev1.SecretTypeSSHAuth).
 		WithData(map[string][]byte{
 			corev1.SSHAuthPrivateKey:  privatePEM,
 			constants.MPISSHPublicKey: ssh.MarshalAuthorizedKey(publicKey),
-		})
-
-	secret.WithOwnerReferences(metav1ac.OwnerReference().
-		WithAPIVersion(trainer.GroupVersion.String()).
-		WithKind(trainer.TrainJobKind).
-		WithName(trainJob.Name).
-		WithUID(trainJob.UID).
-		WithController(true).
-		WithBlockOwnerDeletion(true))
+		}).
+		WithImmutable(true).
+		WithOwnerReferences(metav1ac.OwnerReference().
+			WithAPIVersion(trainer.GroupVersion.String()).
+			WithKind(trainer.TrainJobKind).
+			WithName(trainJob.Name).
+			WithUID(trainJob.UID).
+			WithController(true).
+			WithBlockOwnerDeletion(true)), nil
+}
 
-	return secret, nil
+func sshAuthSecretName(trainJobName string) string {
```

Comment:
Do we really need this util func if is it only used in `buildSSHAuthSecret` ?

Reply:
Nevermind, find that we also use it to get the Secret: https://github.com/kubeflow/trainer/blob/3c9179ea6f2e5327930d41b68aea303338b0793c/pkg/runtime/framework/plugins/mpi/mpi.go#L222